### PR TITLE
Wrap Cylinder Test

### PR DESCRIPTION
**Pull request by pepbos**
This PR contains a new test for the cylinder wrapping algorithm.

It serves as a starting point for a more rigorous testing of the cylinder wrapping, but currently it contains two simple test cases: One in which the start and end points are inside the cylinder (no-wrapping), and the other where the points are perpendicular to the cylinder (have no axial components) resulting in a simple wrapping:

testWrapCylinder-perpendicular-unconstrained-quarter

The resulting wrapped path points obtained from the simulation are compared against the expected answer, given a tolerance. The expected answer was obtained from the solution as is now. The tolerance was chosen arbitrarily, and might change as more tests are added.

The next improvements I planned to include (in different PRs) are:
- [ ] Assert wrapping direction.
- [ ] Assert geodesic path properties, such as the path length, and gradient.
- [ ] Add more test cases,
- [ ] Add permutations to the test cases, such as rotating the scene, or reversing the start and end points.
- [ ] Merging with the cylinder wrapping test currently in `testWrappingAlgorithm.cpp`.

### Brief summary of changes

A basic test for cylinder wrapping is added: `testWrapCylinder.cpp`

### Looking for feedback on...

More ideas on what to test for are welcome!

### CHANGELOG.md (choose one)

- no need to update because...
- updated.

- - -
This change is [Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3498)

### CODE REVIEW

**Comment by pepbos** [comment]
Hi! @adamkewley @nickbianco @aymanhab , I thought I would start with a simple test, and expand from that. Let me know what you think!

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+#include <sstream>
+#include <functional>
+#include <memory>
+
+using namespace OpenSim;
+
+namespace {
+
+    // A path segment determined in terms of the start and end point.
+    struct PathSegment final {
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:38]: Initialize these the members to known values.

Simbody has an "interesting" behavior, which is that the empty constructor (e.g. `SimTK::Vec3 v;`, `()`, or `{}`) will either produce NaNs, or uninitialized garbage, based on compiler flags. See:

https://github.com/simbody/simbody/blob/master/SimTKcommon/SmallMatrix/include/SimTKcommon/internal/Vec.h#L425C5-L425C5

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+            "WrapTestResult{" <<
+            "path: " << result.path << ", " <<
+            "length: " << result.length << "}";
+    }
+
+    // Struct holding the tolerances when asserting the wrapping result.
+    struct WrappingTolerances final {
+        WrappingTolerances() = default;
+
+        WrappingTolerances(double eps):
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:90]: Space before colon

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        double position = 1e-9;
+        double length = 1e-9;
+    };
+
+    bool IsEqual(
+        double lhs,
+        double rhs,
+        double tolerance)
+    {
+        return std::abs(lhs - rhs) < tolerance;
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:104]: Maybe `<=` (e.g. is something being off by exactly one tolerance interval permitted)

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        WrappingTolerances(double eps):
+            position(eps),
+            length(eps)
+        {}
+
+        double position = 1e-9;
+        double length = 1e-9;
+    };
+
+    bool IsEqual(
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:99]: Pedantic: `IsEqual` -> `IsWithinTolerance`/`IsEqualToWithin` etc.

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    struct WrapInput final {
+
+        SimTK::Rotation cylinderOrientation() const {
+            SimTK::Rotation orientation;
+            orientation.setRotationToBodyFixedXYZ(eulerRotations);
+            return orientation;
+        }
+
+        // Endpoints of the total path:
+        PathSegment path = {{}, {}};
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:159]: `SimTK::Vec3`'s initialization rules imply that this defaults to "junk" in release mode (I believe?)

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+            "quadrant: " << input.quadrant << ", " <<
+            "eulerRotations: " << input.eulerRotations << "}";
+    }
+
+    // Simulates the wrapping scenario as configured by the WrapInput.
+    WrapTestResult solve(const WrapInput& input, bool visualize) {
+        Model model;
+        model.setName("testWrapCylinderModel");
+
+        std::unique_ptr<WrapCylinder> cylinder(new WrapCylinder());
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:183]: If it doesn't make indentation too ugly, consider a sub-scope that starts before `cylinder` and ends after `cylinder.release()` so that it is logically impossible for code further down to use the released (null) pointers.

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        model.setUseVisualizer(visualize);
+
+        const SimTK::State& state = model.initSystem();
+
+        if (visualize) {
+            model.realizeVelocity(state);
+            model.getVisualizer().show(state);
+        }
+
+        model.updComponent<PathSpring>("spring").getLength(state);
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:215]: Unclear why this is necessary (presumably, you are using it to force some kind of internal caching/realization)?

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+        if (!IsEqual(result, expected, tol.position)) {
+            std::ostringstream oss;
+            oss << "\nFAILED: case = " << testCase;
+            oss << "\n    input = " << input;
+            oss << "\n    result = " << result;
+            oss << "\n    expected = " << expected;
+            oss << "\n    caused by: "
+                << "Expected and simulated result not equal to within tolerance = "
+                << tol.position;
+            return oss.str();
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:258]: Not a review comment, but understanding overload `(3)` in:

- https://en.cppreference.com/w/cpp/io/basic_ostringstream/str

Is an interesting exercise.

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+        }
+
+        return std::string{};
+    }
+
+}
+
+int main()
+{
+    std::string failLog;
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:268]: `std::vector<std::string>`?

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    failLog += TestWrapping(input, expected, tolerance, name);
+
+    // =========================================================================
+    // ======================= Perpendicular Case ==============================
+    // =========================================================================
+
+    name = "Perpendicular";
+    input.path = {{2, -2, 0}, {-2, 2.1, 0}};
+
+    expected.path = {
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:294]: Where are these numbers from?

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 3 above)
Comment (pepbos) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:111]: makes sense. To pass tolerance of zero :)

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 5 above)
Comment (pepbos) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:166]: Yeah looks like that is the case

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 6 above)
Comment (pepbos) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:190]: Aaah, that's what I was looking for!

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 7 above)
Comment (pepbos) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:222]: yeah the idea was to trigger the realization. Bit strange that `getLength(state)` is marked as `const` though, not sure if that is common practice?

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 10 above)
Comment (pepbos) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:301]: My idea was to add more assertions for each case, such that (hopefully) all edge cases and interesting wrapping conditions can be captured. But before adding those cases, I would like to add the testing of certain geometric properties, such as path gradient, wrapping direction etc. So I just added a simple case for which the wrapping path is very accurate, but is otherwise not necessarily interesting.

But yeah, at this moment, there is no way of testing if this wrapping is accurate, so it seems a bit random.

Comment (adamkewley) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:301]: A comment along the lines of "these were scooped from OpenSim 4.5", "calculated by hand", etc. is enough here

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+    }
+
+    // Simulates the wrapping scenario as configured by the WrapInput.
+    WrapTestResult solve(const WrapInput& input, bool visualize) {
+        Model model;
+        model.setName("testWrapCylinderModel");
+
+        // Add a spring to create the wrapping path.
+        {
+            std::unique_ptr<PathSpring> spring (
```
Comment (nickbianco) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:194]: Remove space after `spring`.

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 7 above)
Comment (nickbianco) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:233]: The `const`-ness of a method doesn't apply to `State` realization. It's related to whether or not a method will change the properties of a OpenSim object (e.g., `upd_stiffness` is non-`const` since it modifies the `PathSpring` stiffness).

**Review by nickbianco — COMMENTED** [review]
Great addition @pepbos! I added a few comments.

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+        PathSegment() = default;
+
+        PathSegment(
+            const SimTK::Vec3& startPoint,
+            const SimTK::Vec3& endPoint) :
+            start(startPoint),
+            end(endPoint)
+        {}
+
+        SimTK::Vec3 start {SimTK::NaN};
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:48]: Initializers don't usually have spaces between the name and the initializing statement (e.g. `std::vector<int> vals(2);`, `std::vector<int> vals{1,2,3};`, not `std::vector<int> vals (2);`)

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+        static WrapTestResult NoWrap() {
+            WrapTestResult result;
+            result.noWrap = true;
+            return result;
+        }
+
+        // Path segment on cylinder surface.
+        PathSegment path;
+        // Path segment length.
+        double length = NAN;
```
Comment (adamkewley) [inline OpenSim/Tests/Wrapping/testWrapCylinder.cpp:77]: Use `NAN`/`SimTK::NaN` consistently - if possible

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 7 above)
Comment (adamkewley) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:233]: It might be useful to document why it's there or, if this works, explicitly realize the state rather than realizing it via this method (e.g. `model.realizeState(state, SimTK::Stage::Whatever);`)

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 9 above)
Comment (adamkewley) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp:286]: Yes, because it makes it explicit in the source code that the variable is "a sequence of error messages" rather than relying on an in-string encoding.

File: OpenSim/Tests/Wrapping/testWrapCylinder.cpp (hunk 7 above)
Comment (pepbos) [reply OpenSim/Tests/Wrapping/testWrapCylinder.cpp]: "Realizing" the state will not trigger the computation, hence this sort of manual triggering of the wrap computation. But I added a comment.

**Comment by pepbos** [comment]
@adamkewley @nickbianco thanks for all your comments! Mostly the resolutions seemed straightforward, but let me know!

**Review by nickbianco — APPROVED** [review]
LGTM!
